### PR TITLE
Language mode utils

### PR DESCRIPTION
**Pull request by peterschrammel**
Document and extend language/mode utility functions as preparation for https://github.com/diffblue/cbmc/pull/2098

### CODE REVIEW

File: src/langapi/mode.cpp [hunk 1]
```diff
@@ -28,6 +31,10 @@ struct language_entryt
 typedef std::list<language_entryt> languagest;
 languagest languages;
 
+/// Register a language
+/// Note: registering a language is required for using the functions
+///   in language_util.h
+/// \param factory: a langugae factory, e.g. `new_ansi_c_language`
```
Comment (tautschnig) [inline src/langapi/mode.cpp]: Should this go in the second commit?

File: src/langapi/mode.cpp [hunk 2]
```diff
@@ -37,17 +44,60 @@ void register_language(language_factoryt factory)
   languages.back().mode=l->id();
 }
 
+/// Get the language corresponding to the given mode
+/// \param mode: the mode, e.g. `ID_C`
+/// \return the language or `nullptr` if the language has not been registered
```
Comment (tautschnig) [inline src/langapi/mode.cpp:49]: Move to second commit?

File: src/langapi/mode.cpp [hunk 3]
```diff
@@ -37,17 +44,60 @@ void register_language(language_factoryt factory)
...
+    language,
+    "symbol `" + id2string(identifier) + "' has unknown mode '" +
+      id2string(mode) + "'");
+  return language;
+}
+
+/// Get the language corresponding to the registered file name extensions
+/// \param filename: a filename
+/// \return the corresponding language or `nullptr` if the extension cannot
+///   be resolved to any registered language
```
Comment (tautschnig) [inline src/langapi/mode.cpp:100]: Move to second commit?

**Review by tautschnig — APPROVED** [review]
LGTM. I'm confused by the Travis error. And I'm wondering whether to place all additional documentation in just one commit.

**Comment by peterschrammel** [comment]
>  I'm confused by the Travis error.

`develop` was only passing because we were always lucky that building `langapi.a` finished before `smt2_solver`. The langapi dependency will finally disappear with #2098.

**Review by smowton — APPROVED** [review]

**Pull request merged**